Revert the following:

commit d9715adf895e9acfbaf17ae05b18f2b2467ca322
Author: Frank Tang <ftang@chromium.org>
Date:   Tue Jun 20 15:12:47 2023 -0700

    [Intl] rm alt_calendar hack post ICU 73-1
    
    Also need to bump up the required ICU version.
    
    Bug: v8:14086
    Change-Id: I52a53fcd201f3272aa712123fc00c54d0b762d53
    Reviewed-on: https://chromium-review.googlesource.com/c/v8/v8/+/4628805
    Reviewed-by: Shu-yu Guo <syg@chromium.org>
    Commit-Queue: Frank Tang <ftang@chromium.org>
    Cr-Commit-Position: refs/heads/main@{#88431}

--- a/v8/src/objects/intl-objects.h
+++ b/v8/src/objects/intl-objects.h
@@ -22,7 +22,7 @@
 #error Internationalization is expected to be enabled.
 #endif  // V8_INTL_SUPPORT
 
-#define V8_MINIMUM_ICU_VERSION 73
+#define V8_MINIMUM_ICU_VERSION 71
 
 namespace U_ICU_NAMESPACE {
 class BreakIterator;
--- a/v8/src/objects/js-date-time-format-inl.h
+++ b/v8/src/objects/js-date-time-format-inl.h
@@ -30,6 +30,8 @@ ACCESSORS(JSDateTimeFormat, icu_date_int
           Tagged<Managed<icu::DateIntervalFormat>>,
           kIcuDateIntervalFormatOffset)
 
+BOOL_ACCESSORS(JSDateTimeFormat, flags, alt_calendar, AltCalendarBit::kShift)
+
 inline void JSDateTimeFormat::set_hour_cycle(HourCycle hour_cycle) {
   int hints = flags();
   hints = HourCycleBits::update(hints, hour_cycle);
--- a/v8/src/objects/js-date-time-format.cc
+++ b/v8/src/objects/js-date-time-format.cc
@@ -595,7 +595,8 @@ Handle<Object> JSDateTimeFormat::TimeZon
 
 namespace {
 Handle<String> GetCalendar(Isolate* isolate,
-                           const icu::SimpleDateFormat& simple_date_format) {
+                           const icu::SimpleDateFormat& simple_date_format,
+                           bool is_alt_calendar = false) {
   // getType() returns legacy calendar type name instead of LDML/BCP47 calendar
   // key values. intl.js maps them to BCP47 values for key "ca".
   // TODO(jshin): Consider doing it here, instead.
@@ -606,9 +607,17 @@ Handle<String> GetCalendar(Isolate* isol
   // and
   // http://www.unicode.org/repos/cldr/tags/latest/common/bcp47/calendar.xml
   if (calendar_str == "gregorian") {
-    calendar_str = "gregory";
+    if (is_alt_calendar) {
+      calendar_str = "iso8601";
+    } else {
+      calendar_str = "gregory";
+    }
   } else if (calendar_str == "ethiopic-amete-alem") {
     calendar_str = "ethioaa";
+  } else if (calendar_str == "islamic") {
+    if (is_alt_calendar) {
+      calendar_str = "islamic-rgsa";
+    }
   }
   return isolate->factory()->NewStringFromAsciiChecked(calendar_str.c_str());
 }
@@ -623,7 +632,8 @@ Handle<Object> GetTimeZone(Isolate* isol
 Handle<String> JSDateTimeFormat::Calendar(
     Isolate* isolate, DirectHandle<JSDateTimeFormat> date_time_format) {
   return GetCalendar(isolate,
-                     *(date_time_format->icu_simple_date_format()->raw()));
+                     *(date_time_format->icu_simple_date_format()->raw()),
+                     date_time_format->alt_calendar());
 }
 
 Handle<Object> JSDateTimeFormat::TimeZone(
@@ -2285,6 +2295,9 @@ MaybeHandle<JSDateTimeFormat> JSDateTime
     icu_locale.setUnicodeKeywordValue("ca", calendar_str.get(), status);
     DCHECK(U_SUCCESS(status));
   }
+  bool alt_calendar =
+      strstr(icu_locale.getName(), "calendar=iso8601") != nullptr ||
+      strstr(icu_locale.getName(), "calendar=islamic-rgsa") != nullptr;
 
   if (numbering_system_str != nullptr &&
       Intl::IsValidNumberingSystem(numbering_system_str.get())) {
@@ -2650,6 +2663,7 @@ MaybeHandle<JSDateTimeFormat> JSDateTime
     date_time_format->set_time_style(time_style);
   }
   date_time_format->set_hour_cycle(dateTimeFormatHourCycle);
+  date_time_format->set_alt_calendar(alt_calendar);
   date_time_format->set_locale(*locale_str);
   date_time_format->set_icu_locale(*managed_locale);
   date_time_format->set_icu_simple_date_format(*managed_format);
@@ -2751,8 +2765,10 @@ MaybeHandle<JSArray> FormatToPartsWithTe
   DateTimeValueRecord x_record;
   MAYBE_ASSIGN_RETURN_ON_EXCEPTION_VALUE(
       isolate, x_record,
-      HandleDateTimeValue(isolate, *format, GetCalendar(isolate, *format), x,
-                          method_name),
+      HandleDateTimeValue(
+          isolate, *format,
+          GetCalendar(isolate, *format, date_time_format->alt_calendar()), x,
+          method_name),
       Handle<JSArray>());
 
   return FormatMillisecondsByKindToArray(isolate, *format, x_record.kind,
@@ -3070,8 +3086,8 @@ MaybeHandle<T> FormatRangeCommonWithTemp
   // 6. Let x be ? HandleDateTimeValue(dateTimeFormat, x).
   icu::SimpleDateFormat* icu_simple_date_format =
       date_time_format->icu_simple_date_format()->raw();
-  Handle<String> date_time_format_calendar =
-      GetCalendar(isolate, *icu_simple_date_format);
+  Handle<String> date_time_format_calendar = GetCalendar(
+      isolate, *icu_simple_date_format, date_time_format->alt_calendar());
   DateTimeValueRecord x_record;
   MAYBE_ASSIGN_RETURN_ON_EXCEPTION_VALUE(
       isolate, x_record,
--- a/v8/src/objects/js-date-time-format.h
+++ b/v8/src/objects/js-date-time-format.h
@@ -155,6 +155,8 @@ class JSDateTimeFormat
   DECL_ACCESSORS(icu_date_interval_format,
                  Tagged<Managed<icu::DateIntervalFormat>>)
 
+  DECL_BOOLEAN_ACCESSORS(alt_calendar)
+
   DECL_PRINTER(JSDateTimeFormat)
 
   TQ_OBJECT_CONSTRUCTORS(JSDateTimeFormat)
--- a/v8/src/objects/js-date-time-format.tq
+++ b/v8/src/objects/js-date-time-format.tq
@@ -11,6 +11,14 @@ bitfield struct JSDateTimeFormatFlags ex
   hour_cycle: HourCycle: 3 bit;
   date_style: DateTimeStyle: 3 bit;
   time_style: DateTimeStyle: 3 bit;
+  // ICU report the same type "gregorian" for both "gregorian" calendar and
+  // "iso8601" calendar and the same type "islamic" for both "islamic" and
+  // "islamic-rgsa" calendar. We use the alt_calendar bit to distinguish between
+  // them. When the type is "gregorian" and the alt_calendar bit is set, it is
+  // "iso8601", otherwise the true "gregorian" calendar. While the type is
+  // "islamic" and the alt_calendar bit is set, it is "islamic-rgsa" calendar,
+  // otherwise "islamic" calendar.
+  alt_calendar: bool: 1bit;
 }
 
 extern class JSDateTimeFormat extends JSObject {
